cw_zano/lib/zano_wallet.dart

```diff
@@ -0,0 +1,533 @@
+            privateViewKey: walletInfo.wiExtended.viewPrivateKey,
+            publicSpendKey: walletInfo.wiExtended.spendPublicKey,
+            publicViewKey: walletInfo.wiExtended.viewPublicKey,
+          );
+          loadAssets(walletInfo.wi.balances);
+          // matching balances and whitelists
+          // 1. show only balances available in whitelists
+          // 2. set whitelists available in balances as 'enabled' ('disabled' by default)
+          for (final b in walletInfo.wi.balances) {
+            if (b.assetId == zanoAssetId) {
+              balance[CryptoCurrency.zano] = ZanoBalance(total: b.total, unlocked: b.unlocked);
+            } else {
+              final asset = zanoAssets[b.assetId];
+              if (asset == null) {
+                ZanoWalletApi.error('balance for an unknown asset ${b.assetInfo.assetId}');
+                continue;
+              }
+              if (balance.keys.any((element) => element is ZanoAsset && element.assetId == b.assetInfo.assetId)) {
+                balance[balance.keys.firstWhere((element) => element is ZanoAsset && element.assetId == b.assetInfo.assetId)] =
+                    ZanoBalance(total: b.total, unlocked: b.unlocked, decimalPoint: asset.decimalPoint);
+              } else {
+                balance[asset] = ZanoBalance(total: b.total, unlocked: b.unlocked, decimalPoint: asset.decimalPoint);
+              }
+            }
+          }
+          await updateTransactions(); 
+          // removing balances for assets missing in wallet info balances
+          balance.removeWhere(
+            (key, _) => key != CryptoCurrency.zano && !walletInfo.wi.balances.any((element) => element.assetId == (key as ZanoAsset).assetId),
+          );
+        }
+      });
+    } catch (e) {
+      syncStatus = FailedSyncStatus();
+      ZanoWalletApi.error(e.toString());
+      //rethrow; // TODO: we don't need to propagate exception here 
+    }
+  }
+
+  @override
+  Future<void>? updateBalance() => null;
+
+  Future<void> updateTransactions() async {
+    try {
+      if (_isTransactionUpdating) {
+        return;
+      }
+      _isTransactionUpdating = true;
+      final transactions = await fetchTransactions();
+      if (transactions.length == transactionHistory.transactions.length) {
+        return;
+      }
+      transactionHistory.clear();
+      transactionHistory.addMany(transactions);
+      await transactionHistory.save();
+      _isTransactionUpdating = false;
+    } catch (e) {
+      printV("e: $e");
+      ZanoWalletApi.error(e.toString());
+      _isTransactionUpdating = false;
+    }
+  }
+
+  Future<CryptoCurrency> addZanoAssetById(String assetId) async {
+    if (zanoAssets.containsKey(assetId)) {
+      throw ZanoWalletException('zano asset with id $assetId already added');
+    }
+    final assetDescriptor = await addAssetsWhitelist(assetId);
+    if (assetDescriptor == null) {
+      throw ZanoWalletException("there's no zano asset with id $assetId");
```

Comment:
let's make sure that those exceptions are caught and handled gracefully instead of showing the report error popup

Reply:
added error handling in edit_token_page.dart